### PR TITLE
Unpin mpi4py version

### PR DESCRIPTION
**Pull request by JDBetteridge**
# Description

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -83,16 +83,20 @@ jobs:
             --install defcon \
             --install gadopt \
             --install asQ \
+            --package-branch petsc JDBetteridge/merge_upstream \
+            --package-branch slepc JDBetteridge/merge_upstream \
```
Comment (JDBetteridge) [inline .github/workflows/build.yml:87]:
```suggestion
            --package-branch slepc JDBetteridge/merge-upstream \
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -83,16 +83,20 @@ jobs:
             --install defcon \
             --install gadopt \
             --install asQ \
+            --package-branch petsc JDBetteridge/merge_upstream \
```
Comment (JDBetteridge) [inline .github/workflows/build.yml:86]:
```suggestion
            --package-branch petsc JDBetteridge/fix_mpi4_build \
```

**Review by dham — CHANGES_REQUESTED** [review]
Looks good except that we should enable a `--mpi4py-version=3.1.6` option to the install script and have it honour that on lines 1859 and 2004 so that we can bail out users of older OSes who need to install with mpi4py v3

File: scripts/firedrake-install [hunk 3]
```diff
@@ -442,6 +443,7 @@ else:
                         help="Specify path to system BLAS directory. Use '--with-blas=download' to download openblas")
     parser.add_argument("--netgen", action="store_true", dest="netgen", default=config["options"].get("netgen", False),
                         help="Install Netgen.")
+    parser.add_argument("--mpi4py-version", help="Specify an exact version of mpi4py to install")
```
Comment (JHopeCollins) [inline scripts/firedrake-install:446]: Did you mean to add this argument twice?

**Pull request merged**